Do not write retriable errors for Replicated mutate/merge into error log

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not write retriable errors for Replicated mutate/merge into error log (fixes tons of `No active replica has part X or covering part` in the error log, since they are not actually errors)

Cc: @tavplubix 
Fixes: #50395 (cc @CheSema )
Fixes: e3f892f ("fix gtest with MemoryWriteBuffer, do not mute exception in ReplicatedMergeMutateTaskBase")

_Note: marked as bug fix, since it pollute logs_

azat committed:
Do not write retriable errors for Replicated mutate/merge into error log

Fixes: e3f892f84a1 ("fix gtest with MemoryWriteBuffer, do not mute exception in ReplicatedMergeMutateTaskBase")
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add a test for retrible errors (i.e. "No active replica has part {} or covering part")

They should not appears in the error log, only with Information level.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #55944 from azat/ClickHouse:rmt-retriable-exception-log -->
*This is an automated comment for commit bd26f7096a4a3325f7a363c4be919700cdf10ca3 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55944/bd26f7096a4a3325f7a363c4be919700cdf10ca3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6613827327">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55944/bd26f7096a4a3325f7a363c4be919700cdf10ca3/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp:
```diff
@@ -69,11 +69,11 @@ bool ReplicatedMergeMutateTaskBase::executeStep()
+            /// This exception will be written to the queue element, and it can be looked up using `system.replication_queue` table.
+            /// The thread that performs this action will sleep a few seconds after the exception.
+            /// See `queue.processEntry` function.
+            if (!retryable_error)
+                throw;
```
Unfortunately, it will not work because we have some logic built on top of exceptions...

tavplubix replied:
Btw, I was thinking that maybe we have some backoff on retries of failed queue entries, it would help too

azat replied:
>Unfortunately, it will not work because we have some logic built on top of exceptions... (so we have to rethow it unconditionally, otherwise replication queue may get stuck, as you can see in the failed tests)

Yeah, now I see, though it is not about replication queue, but about abort in the WriteBuffer dtor, but maybe there are more problems?

Anyway, I rewrote the patch, introduced `IExecutableTask::printExecutionException()` to let the `MergeTreeBackgroundExecutor` decide. Looks a little bit hackish I too much, maybe, but what do you think?

azat committed:
Convert 02903_rmt_retriable_merge_exception to an sh test with proper retries

CI: https://s3.amazonaws.com/clickhouse-test-reports/55944/bd26f7096a4a3325f7a363c4be919700cdf10ca3/stateless_tests_flaky_check__asan_.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Proper ignoring of retriable errors during merge/mutate

The exception cannot be simply suppressed, since sanity checks in the
WriteBuffer dtor relies on the std::uncaught_exceptions(), and if the
exception will be suppressed, then if the buffer was not finalized, it
may abort (even though it is only in debug/sanitizers build).

So instead, IExecutableTask::printExecutionException() had been
introduced to distinguish when exception should be printed and when not.

azat committed:
tests: more robust 02903_rmt_retriable_merge_exception

- separate uuid extraction
- add preliminary exit
- disable for ordinary database
- less number of attempts
- add optimize_throw_if_noop and missing sync replica, to fix:

  2023.10.24 15:18:35.925533 [ 640 ] {da7418c6-3d51-45bc-a0d0-4970bb0cdd51} <Debug> test_3kgjgry1.rmt1 (d18afb81-3a4b-4c02-b281-5575dce2f440): Cannot select parts for optimization: Entry for part all_1_1_0 hasn't been read from the replication log yet (in partition all)

- fix in case of ZooKeeper retries

  2023.10.24 11:50:24.792511 [ 1437 ] {c39fd15b-e2e6-4291-9912-39fda75ebcd5} <Trace> test_qxkzmigq.rmt1 (1c086c74-9ebe-495c-bbd2-87ab2d8ec43d): Renaming temporary part tmp_insert_all_1_1_0 to all_1_1_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
  2023.10.24 11:50:24.797320 [ 1437 ] {c39fd15b-e2e6-4291-9912-39fda75ebcd5} <Trace> test_qxkzmigq.rmt1 (1c086c74-9ebe-495c-bbd2-87ab2d8ec43d) (Replicated OutputStream): ZooKeeperWithFaultInjection call FAILED: seed=17644626169032325693 func=tryMulti path=/clickhouse/zero_copy code=Session expired message=Fault injection before operation
  2023.10.24 11:50:24.797536 [ 1437 ] {c39fd15b-e2e6-4291-9912-39fda75ebcd5} <Debug> test_qxkzmigq.rmt1 (1c086c74-9ebe-495c-bbd2-87ab2d8ec43d): Undoing transaction. Rollbacking parts state to temporary and removing from working set: all_1_1_0.
  ...
  2023.10.24 11:50:25.000349 [ 1437 ] {c39fd15b-e2e6-4291-9912-39fda75ebcd5} <Trace> test_qxkzmigq.rmt1 (1c086c74-9ebe-495c-bbd2-87ab2d8ec43d): Renaming temporary part tmp_insert_all_1_1_0 to all_2_2_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
  2023.10.24 11:50:25.007477 [ 760 ] {} <Trace> test_qxkzmigq.rmt1 (ReplicatedMergeTreeQueue): Insert entry queue-0000000000 to queue with type GET_PART with virtual parts [all_2_2_0]

CI:
- https://s3.amazonaws.com/clickhouse-test-reports/55944/da408df4a7296835897d7cef80d63f252df79b75/stateless_tests__tsan__s3_storage__[2_5].html
- https://s3.amazonaws.com/clickhouse-test-reports/55944/da408df4a7296835897d7cef80d63f252df79b75/stateless_tests_flaky_check__asan_.html
- https://s3.amazonaws.com/clickhouse-test-reports/55944/02fdd0513f7d413ce4ac39a70566855327ebfade/stateless_tests__aarch64_.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

tavplubix commented:
Integration tests (release) [4/4] - https://github.com/ClickHouse/ClickHouse/issues/54502
Stateless tests (aarch64) - https://github.com/ClickHouse/ClickHouse/pull/56023
